File: samples/csharp_dotnetcore/42.scaleout/BlobStore.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using Microsoft.WindowsAzure.Storage;
using Microsoft.WindowsAzure.Storage.Auth;
using Microsoft.WindowsAzure.Storage.Blob;
using Newtonsoft.Json.Linq;

namespace Microsoft.BotBuilderSamples
{
    /// <summary>
    /// An implementation of the ETag aware IStore interface against Azure Blob Storage.
    /// </summary>
    public class BlobStore : IStore
    {
        private readonly CloudBlobContainer _container;

        public BlobStore(string accountName, string accountKey, string containerName)
        {
            if (string.IsNullOrWhiteSpace(accountName))
            {
                throw new ArgumentException(nameof(accountName));
            }

            if (string.IsNullOrWhiteSpace(accountKey))
            {
                throw new ArgumentException(nameof(accountKey));
            }

            if (string.IsNullOrWhiteSpace(containerName))
            {
                throw new ArgumentException(nameof(containerName));
            }

            var storageCredentials = new StorageCredentials(accountName, accountKey);
            var cloudStorageAccount = new CloudStorageAccount(storageCredentials, useHttps: true);
            var client = cloudStorageAccount.CreateCloudBlobClient();
            _container = client.GetContainerReference(containerName);
        }

        public async Task<(JObject content, string etag)> LoadAsync(string key)
        {
            if (string.IsNullOrWhiteSpace(key))
            {
                throw new ArgumentException(nameof(key));
            }

            var blob = _container.GetBlockBlobReference(key);
            try
            {
                var content = await blob.DownloadTextAsync();
                var obj = JObject.Parse(content);
                var etag = blob.Properties.ETag;
                return (obj, etag);
            }
            catch (StorageException e)
                when (e.RequestInformation.HttpStatusCode == (int)HttpStatusCode.NotFound)
            {
                return (new JObject(), null);
            }
        }

        public async Task<bool> SaveAsync(string key, JObject obj, string etag)
        {
            if (string.IsNullOrWhiteSpace(key))
            {
                throw new ArgumentException(nameof(key));
            }

            if (obj == null)
            {
                throw new ArgumentNullException(nameof(obj));
            }

            var blob = _container.GetBlockBlobReference(key);
            blob.Properties.ContentType = "application/json";
            var content = obj.ToString();
            if (etag != null)
            {
                try
                {
                    await blob.UploadTextAsync(content, Encoding.UTF8, new AccessCondition {IfMatchETag = etag}, new BlobRequestOptions(), new OperationContext());
                }
                catch (StorageException e)
                    when (e.RequestInformation.HttpStatusCode == (int)HttpStatusCode.PreconditionFailed)
                {
                    return false;
                }
            }
            else
            {
                await blob.UploadTextAsync(content);
            }

            return true;
        }
    }
}

File: samples/javascript_nodejs/42.scale-out/blobStore.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { BlobServiceClient, StorageSharedKeyCredential } = require('@azure/storage-blob');

class BlobStore {
    constructor(accountName, accountKey, containerName) {
        if (!accountName) {
            throw new Error('accountName is required');
        }

        if (!accountKey) {
            throw new Error('accountKey is required');
        }

        if (!containerName) {
            throw new Error('containerName is required');
        }

        const sharedKeyCredential = new StorageSharedKeyCredential(accountName, accountKey);
        const blobServiceClient = new BlobServiceClient(`https://${ accountName }.blob.core.windows.net`, sharedKeyCredential);
        this.containerClient = blobServiceClient.getContainerClient(containerName);
    }

    async loadAsync(key) {
        if (!key) {
            throw new Error('key is required');
        }

        const blobClient = this.containerClient.getBlockBlobClient(key);
        try {
            const downloadBlockBlobResponse = await blobClient.download();
            const content = await streamToString(downloadBlockBlobResponse.readableStreamBody);
            const obj = JSON.parse(content);
            const etag = downloadBlockBlobResponse.properties.etag;
            return { content: obj, etag: etag };
        } catch (error) {
            if (error.statusCode === 404) {
                return { content: {}, etag: null };
            }
            throw error;
        }
    }

    async saveAsync(key, obj, etag) {
        if (!key) {
            throw new Error('key is required');
        }

        if (!obj) {
            throw new Error('obj is required');
        }

        const blobClient = this.containerClient.getBlockBlobClient(key);
        blobClient.properties.contentType = 'application/json';
        const content = JSON.stringify(obj);
        if (etag) {
            try {
                await blobClient.upload(content, content.length, { conditions: { ifMatch: etag } });
            } catch (error) {
                if (error.statusCode === 412) {
                    return false;
                }
                throw error;
            }
        } else {
            await blobClient.upload(content, content.length);
        }

        return true;
    }
}

async function streamToString(readableStream) {
    return new Promise((resolve, reject) => {
        const chunks = [];
        readableStream.on('data', (data) => {
            chunks.push(data.toString());
        });
        readableStream.on('end', () => {
            resolve(chunks.join(''));
        });
        readableStream.on('error', reject);
    });
}

module.exports.BlobStore = BlobStore;
